wrappers/pkcs11/pkcs11_client.go

```diff
@@ -0,0 +1,553 @@
+	case opts.withKeyId != "":
+		keyId = opts.withKeyId
+	default:
+		keyId = ""
+	}
+
+	switch {
+	case api.ReadBaoVariable(EnvHsmWrapperPin) != "" && !opts.Options.WithDisallowEnvVars:
+		pin = api.ReadBaoVariable(EnvHsmWrapperPin)
+	case opts.withPin != "":
+		pin = opts.withPin
+	default:
+		return nil, nil, fmt.Errorf("pin is required")
+	}
+
+	switch {
+	case api.ReadBaoVariable(EnvHsmWrapperKeyLabel) != "" && !opts.Options.WithDisallowEnvVars:
+		keyLabel = api.ReadBaoVariable(EnvHsmWrapperKeyLabel)
+	case opts.withKeyLabel != "":
+		keyLabel = opts.withKeyLabel
+	default:
+		return nil, nil, fmt.Errorf("key label is required")
+	}
+
+	switch {
+	case api.ReadBaoVariable(EnvHsmWrapperMechanism) != "" && !opts.Options.WithDisallowEnvVars:
+		mechanism = api.ReadBaoVariable(EnvHsmWrapperMechanism)
+	case opts.withMechanism != "":
+		mechanism = opts.withMechanism
+	default:
+		mechanism = ""
+	}
+
+	if slot != "" {
+		if slotNum, err = numberAutoParse(slot, 32); err != nil {
+			return nil, nil, fmt.Errorf("Invalid slot number")
+		}
+	} else {
+		slotNum = 0
+	}
+
+	if mechanism != "" {
+		if mechanismNum, err = MechanismFromString(mechanism); err != nil {
+			return nil, nil, err
+		}
+	} else {
+		mechanismNum = 0
+	}
+
+	client := &Pkcs11Client{
+		client:    	nil,
+		lib:		lib,
+		slot:      	uint(slotNum),
+		pin:       	pin,
+		tokenLabel: tokenLabel,
+		keyId:		keyId,
+		keyLabel:  	keyLabel,
+		mechanism: 	uint(mechanismNum),
+	}
+	
+	// Initialize the client
+	_, err = client.GetClient()
+	if err != nil {
+		return nil, nil, err
+	}
+	// Validate credentials for session establishment
+	session, err := client.GetSession()
+	if err != nil {
+		return nil, nil, err
+	}
+	defer client.CloseSession(session)
+
+	wrapConfig := new(wrapping.WrapperConfig)
+	wrapConfig.Metadata = make(map[string]string)
+	wrapConfig.Metadata["lib"] = lib
+	wrapConfig.Metadata["key_label"] = keyLabel
+	wrapConfig.Metadata["key_id"] = keyId
+	if slotNum != 0 {
+		wrapConfig.Metadata["slot"] = strconv.Itoa(int(slotNum))
+	}
+	if tokenLabel != "" {
+		wrapConfig.Metadata["token_label"] = tokenLabel
+	}
+	if mechanismNum != 0 {
+		wrapConfig.Metadata["mechanism"] = MechanismString(uint(mechanismNum))
+	}
+
+	return client, wrapConfig, nil
+}
+
+
+func (c *Pkcs11Client) Close() {
+	if c.client == nil {
+		return
+	}
+	c.client.Finalize()
+	c.client.Destroy()
+	c.client = nil
+}
+
+func (c *Pkcs11Client) Encrypt(plaintext []byte) ([]byte, *Pkcs11Key, error) {
+	session, err := c.GetSession()
+	if err != nil {
+		return nil, nil, err
+	}
+	defer c.CloseSession(session)
+
+	keyId := Pkcs11Key{ label: c.keyLabel, id: c.keyId }
+	obj, err := c.FindKey(session, keyId, pkcs11.CKA_ENCRYPT)
+	if err != nil {
+		return nil, nil, err
+	}
+	if len(obj) != 1 {
+		return nil, nil, fmt.Errorf("expected 1 object, got %d", len(obj))
+	}
+	key := obj[0]
+
+	mechanism, err := c.GetKeyMechanism(session, key)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var iv []byte
```

Comment:
@glatigny I think you need to correctly handle mechanism parameters here. See https://pkg.go.dev/github.com/miekg/pkcs11#NewGCMParams or https://pkg.go.dev/github.com/miekg/pkcs11#NewOAEPParams.

Otherwise, this fails for me:

```
    pkcs11_acc_test.go:32: err : failed to pkcs11 EncryptInit: pkcs11: 0x7: CKR_ARGUMENTS_BAD
```

Reply:
I ended up with a Dockerfile:

```
FROM registry.fedoraproject.org/fedora:latest

COPY . /go-kms-wrapping


# pkcs11-tool --module "$BAO_HSM_LIB" --token-label "$BAO_HSM_TOKEN_LABEL" --so-pin 1234 --pin 4321 --keygen --key-type aes:32 --label "$BAO_HSM_KEY_LABEL" && \

# Set up SoftHSM, generate a key, and run tests.
RUN true && \
    dnf update --refresh -y && \
    dnf install -y softhsm opensc golang && \
    mkdir /softhsm && \
    export SOFTHSM2_CONF=/go-kms-wrapping/wrappers/pkcs11/softhsm2.conf && \
    softhsm2-util --init-token --slot 0 --label "OpenBao g-k-w" --so-pin 1234 --pin 4321 && \
    softhsm2-util --show-slots && \
    export BAO_HSM_TOKEN_LABEL="OpenBao g-k-w" && \
    export BAO_HSM_PIN=4321 && \
    export BAO_HSM_LIB=/usr/lib64/softhsm/libsofthsm.so && \
    export BAO_HSM_KEY_LABEL=bao-root-key && \
    pkcs11-tool --module "$BAO_HSM_LIB" --token-label "$BAO_HSM_TOKEN_LABEL" --so-pin 1234 --pin 4321 --keypairgen --key-type rsa:4096 --label "$BAO_HSM_KEY_LABEL" && \
    pkcs11-tool --module "$BAO_HSM_LIB" --token-label "$BAO_HSM_TOKEN_LABEL" --so-pin 1234 --pin 4321 --show-info && \
    cd /go-kms-wrapping/wrappers/pkcs11 && \
    export VAULT_ACC=true && \
    go test -v ./... && \
    true
```

with config (`wrappers/pkcs11/softhsm2.conf`):

```
directories.tokendir = /softhsm
objectstore.backend = file
log.level = DEBUG
```

and run via:

```
$ podman build -f wrappers/pkcs11/Dockerfile .
```